sdk/servicebus/azservicebus/processor.go

```diff
@@ -0,0 +1,395 @@
+
+	// TODO: `listen` doesn't give you additional messages unless you've settled previous ones.
+	// In track 2 SDKs you should get a new message, regardless of settlement (ie: it's simply gated on the # of outstanding callbacks)
+	listenHandle := receiver.Listen(ctx, internal.HandlerFunc(func(ctx context.Context, legacyMessage *internal.Message) error {
+		// this shouldn't happen since we do a `select` above that prevents it.
+		// errors from their handler are sent to their error handler but do not terminate the
+		// subscription.
+		activeCallbacksWg.Add(1)
+		defer activeCallbacksWg.Done()
+
+		return handleSingleMessage(handleMessage, notifyErrorAsync, shouldAutoComplete, receiver, legacyMessage)
+	}))
+
+	select {
+	case <-ctx.Done():
+		notifyErrorAsync(ctx.Err())
+		activeCallbacksWg.Wait()
+		return !shouldRetry
+	case <-listenHandle.Done(): // TODO: eliminate this redundancy soon, and hopefully we only need to rely on ctx.Done() above.
+		notifyErrorAsync(listenHandle.Err())
+		activeCallbacksWg.Wait()
+
+		// we should retry since the listen handle can be closed if we did a .Recover() on the receiver.
+		return shouldRetry
+	}
+}
+
+func handleSingleMessage(handleMessage func(message *ReceivedMessage) error, notifyErrorAsync func(err error), shouldAutoComplete bool, receiver internal.LegacyReceiver, legacyMessage *internal.Message) error {
+	err := handleMessage(convertToReceivedMessage(legacyMessage))
+
+	if err != nil {
+		notifyErrorAsync(err)
+	}
+
+	var settleErr error
+
+	if shouldAutoComplete {
+		// NOTE: we ignore the passed in context. Since we're settling behind the scenes
+		// it's nice to wrap it up so users don't have to track it.
+		if err != nil {
+			settleErr = receiver.AbandonMessage(context.Background(), legacyMessage)
+		} else {
+			settleErr = receiver.CompleteMessage(context.Background(), legacyMessage)
+		}
+
+		if settleErr != nil {
+			notifyErrorAsync(settleErr)
+		}
+	}
+
+	return nil
+}
+
+//
+// settlement methods
+// TODO: in other processor implementations this is implemented in the argument for the processMessage
+// callback. You need some sort of association or else you have to track message <-> receiver mappings.
+//
+
+func (p *Processor) CompleteMessage(ctx context.Context, message *ReceivedMessage) error {
+	return message.legacyMessage.Complete(ctx)
+}
+
+func (p *Processor) DeadLetterMessage(ctx context.Context, message *ReceivedMessage) error {
+	// TODO: expand to let them set the reason and description.
+	return message.legacyMessage.DeadLetter(ctx, nil)
+}
+
+func (p *Processor) AbandonMessage(ctx context.Context, message *ReceivedMessage) error {
+	return message.legacyMessage.Abandon(ctx)
+}
+
+func (p *Processor) DeferMessage(ctx context.Context, message *ReceivedMessage) error {
+	return message.legacyMessage.Defer(ctx)
+}
+
+func formatEntity(entity struct {
+	Subqueue     SubQueue
+	Queue        string
+	Topic        string
+	Subscription string
+}) (string, error) {
```

Comment:
Curiosity: why not have this struct defined outside of the method signature?

Reply:
Just trying to avoid creating excess types when they're only used once, or in a really limited scope..